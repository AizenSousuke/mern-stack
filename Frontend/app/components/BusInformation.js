import React, { useState } from "react";
import { View } from "react-native";
import {
	Button,
	ButtonGroup,
	Card,
	Header,
	Overlay,
	TabView,
	Text,
} from "react-native-elements";
import { SafeAreaView } from "react-native-safe-area-context";
import AppStyles from "../../assets/css/AppStyles";
import ColourScheme from "../settings/ColourScheme.json";

const BusInformation = ({ busNumber }) => {
	const [selectedIndex, setSelectedIndex] = useState(0);
	return (
		<SafeAreaView>
			<Header
				centerComponent={
					<Text style={AppStyles.busRouteNumber}>
						Bus {busNumber}
					</Text>
				}
				backgroundColor={ColourScheme.header}
			/>
			<ButtonGroup
				containerStyle={{ marginBottom: 0 }}
				selectedButtonStyle={AppStyles.buttonGroupStyle}
				buttons={["Information", "Route"]}
				selectedIndex={selectedIndex}
				onPress={(index) => {
					setSelectedIndex(index);
				}}
			/>
			{selectedIndex === 0 && (
				<Card>
					<Card.Title>Information</Card.Title>
					<Card.Divider width={1} />
				</Card>
			)}
			{selectedIndex === 1 && (
				<Card>
					<Card.Title>Route</Card.Title>
					<Card.Divider width={1} />
				</Card>
			)}
		</SafeAreaView>
	);
};

export default BusInformation;
