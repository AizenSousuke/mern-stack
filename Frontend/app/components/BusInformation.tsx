import React, { useState, useEffect } from "react";
import { ScrollView, View } from "react-native";
import { ButtonGroup, Card, Chip, Header, Icon, Text } from "react-native-elements";
import { SafeAreaView } from "react-native-safe-area-context";
import AppStyles from "../../assets/css/AppStyles";
import ColourScheme from "../settings/ColourScheme.json";
import { GetBusData, GetBusRouteData } from "../api/api";
import Table from "./Table";

const BusInformation = ({
	busNumber,
	busStopCode,
}: {
	busNumber: string;
	busStopCode: string;
}) => {
	const [selectedIndex, setSelectedIndex] = useState(0);
	const [information, setInformation] = useState([{}]);
	const [route, setRoute] = useState([{}]);

	useEffect(() => {
		(async () => {
			await updatePageData(selectedIndex);
		})();
	}, []);

	const updatePageData = async (index: number) => {
		var data = null;
		setSelectedIndex(index);
		switch (index) {
			case 1:
				// Bus Routes
				console.log("Getting bus route data for bus number", busNumber);
				data = await GetBusRouteData(busNumber);
				if (data) {
					setRoute(data.routes);
				}
				break;

			default:
				// Bus Data
				console.log(
					"Getting bus data for bus number",
					busNumber,
					"and bus stop code",
					busStopCode
				);
				data = await GetBusData(busNumber, busStopCode);
				if (data) {
					setInformation(data.routes);
				}
				break;
		}
	};

	return (
		<SafeAreaView>
			<Header
				centerComponent={
					<Text style={AppStyles.busRouteNumber}>
						Bus {busNumber}
					</Text>
				}
				backgroundColor={ColourScheme.header}
			/>
			<ButtonGroup
				containerStyle={{ marginBottom: 0 }}
				selectedButtonStyle={AppStyles.buttonGroupStyle}
				buttons={["Information", "Route"]}
				selectedIndex={selectedIndex}
				onPress={(index) => {
					updatePageData(index);
				}}
			/>
			<ScrollView style={{ flexGrow: 0 }}>
				{selectedIndex === 0 && (
					<Card>
						<Card.Title>Information</Card.Title>
						<Card.Divider width={1} />
						<Table information={information} />
					</Card>
				)}
				{selectedIndex === 1 && (
					<Card>
						<Card.Title>Route</Card.Title>
						<Card.Divider width={1} />
						{/* <StepIndicator
							direction="vertical"
							currentPosition={1}
							stepCount={route.length}
							labels={route.map((r) => r.BusStopCode)}
							// stepCount={3}
							// labels={["Step 1", "Step 2", "Step 3"]}
							// renderStepIndicator={(position, stepStatus) => {
							// 	<Text key={position}>Indicator: {position}</Text>
							// }}
							// renderLabel={(position, stepStatus, label) => {
							// 	<Text key={position}>Label: {position}{stepStatus}{label}</Text>
							// }}
						/> */}
						{route.map((r, index) => {
							return (
								<View key={index}>
									<Chip
										key={index}
										title={
											r.BusStopData[0].Description +
											" (" +
											r.BusStopCode +
											")"
										}
										onPress={() => {}}
									></Chip>
									{index != route.length - 1 ? <Icon name="caret-down" type="font-awesome" /> : <></>}
									
								</View>
							);
						})}
					</Card>
				)}
			</ScrollView>
		</SafeAreaView>
	);
};

export default BusInformation;
