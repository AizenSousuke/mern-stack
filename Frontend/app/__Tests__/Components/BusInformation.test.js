import BusInformation from "../../components/BusInformation";
import { fireEvent, render, waitFor } from "@testing-library/react-native";

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");

describe("BusInformation", () => {
	it("renders correctly", () => {
		render(<BusInformation busNumber={"966"} busStopCode={"44221"} />);
	});

	it("renders according to snapshot", () => {
		const snapshot = render(
			<BusInformation busNumber={"966"} busStopCode={"44221"} />
		).toJSON();
		expect(snapshot).toMatchSnapshot();
	});

    // Unmounted component error
	// it("should show routes page when tab is switched", () => {
	// 	const screen = render(
	// 		<BusInformation busNumber={"966"} busStopCode={"44221"} />
	// 	);
	// 	const routeTab = screen.getByTestId("RouteButton");
	// 	fireEvent.press(routeTab);
	// 	waitFor(() => {
	// 		const routePage = screen.getByTestId("RoutePage");
    //         expect(routePage).toBeVisible();
	// 	});
	// });
});
