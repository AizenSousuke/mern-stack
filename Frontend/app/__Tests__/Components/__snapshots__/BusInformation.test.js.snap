// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusInformation renders according to snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#DC2F02",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="headerContainer"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          style={
            {
              "alignSelf": "center",
              "color": "white",
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Bus 
          966
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
  <View
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#e3e3e3",
        "borderRadius": 3,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 40,
        "marginBottom": 0,
        "marginHorizontal": 10,
        "marginVertical": 5,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        {
          "borderRightColor": "#bdc6cf",
          "borderRightWidth": 1,
          "flex": 1,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "flex": 1,
            "opacity": 1,
          }
        }
        testID="buttonGroupItem"
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#370617",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 16.714285714285715,
                "fontWeight": "500",
              }
            }
            testID="buttonGroupItemText"
          >
            <View
              style={
                [
                  {
                    "borderRadius": 3,
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 3,
                  },
                  undefined,
                  false,
                ]
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": false,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#2089dc",
                      "borderColor": "#2089dc",
                      "borderRadius": 3,
                      "borderWidth": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "padding": 8,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "white",
                        "fontSize": 18,
                        "paddingVertical": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    Information
                  </Text>
                </View>
              </View>
            </View>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "flex": 1,
            "opacity": 1,
          }
        }
        testID="buttonGroupItem"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "#5e6977",
                "fontSize": 16.714285714285715,
                "fontWeight": "500",
              }
            }
            testID="buttonGroupItemText"
          >
            <View
              style={
                [
                  {
                    "borderRadius": 3,
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 3,
                  },
                  undefined,
                  false,
                ]
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": false,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="RouteButton"
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#2089dc",
                      "borderColor": "#2089dc",
                      "borderRadius": 3,
                      "borderWidth": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "padding": 8,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "white",
                        "fontSize": 18,
                        "paddingVertical": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    Route
                  </Text>
                </View>
              </View>
            </View>
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    style={
      {
        "flexGrow": 0,
      }
    }
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#e1e8ee",
            "borderWidth": 1,
            "margin": 15,
            "marginBottom": 0,
            "padding": 15,
            "shadowColor": "rgba(0,0,0, .2)",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
            }
          }
        >
          <Text
            style={
              {
                "color": "#43484d",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 15,
                "textAlign": "center",
              }
            }
            testID="cardTitle"
          >
            Information
          </Text>
          <View
            style={
              {
                "borderBottomColor": "#bcbbc1",
                "borderBottomWidth": 1,
                "marginBottom": 15,
              }
            }
          />
          <View
            style={
              {
                "display": "flex",
              }
            }
          >
            <View
              style={
                {
                  "display": "flex",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    },
                    {
                      "backgroundColor": "#ffbaba",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#242424",
                        "fontSize": 12,
                      }
                    }
                  >
                    First Bus
                  </Text>
                </View>
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#242424",
                        "fontSize": 12,
                      }
                    }
                  >
                    Last Bus
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      WEEKDAY
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      WEEKDAY
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
              </View>
              <View
                style={
                  {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      SAT
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      SAT
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
              </View>
              <View
                style={
                  {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      SUN
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
                <View
                  style={
                    {
                      "borderColor": "#ffbaba",
                      "borderWidth": 1,
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                      SUN
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flex": 0.5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#242424",
                          "fontSize": 12,
                        }
                      }
                    >
                       hrs
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
